src/gt4py/next/program_processors/runners/dace_fieldview/gtir_sdfg.py

```diff
@@ -172,60 +201,73 @@ def _make_array_shape_and_strides(
 
         Returns:
-            List of data containers or symbols allocated as storage. This is a list, not a single value,
-            because in case of tuples we flat the tuple fields (eventually nested) and allocate storage
-            for each tuple element.
+            List of data containers or symbols allocated as storage. This is a list,
+            not a single value, because in case of tuples we flat the tuple fields
+            (eventually nested) and allocate storage for each tuple element.
         """
-        if isinstance(symbol_type, ts.TupleType):
+        if isinstance(gt_type, ts.TupleType):
             tuple_fields = []
             for tname, tsymbol_type in dace_gtir_utils.get_tuple_fields(
-                name, symbol_type, flatten=True
+                name, gt_type, flatten=True
             ):
                 tuple_fields.extend(
-                    self._add_storage(sdfg, tname, tsymbol_type, transient, is_tuple_member=True)
+                    self._add_storage(sdfg, known_symbols, tname, tsymbol_type, transient)
                 )
             return tuple_fields
 
-        elif isinstance(symbol_type, ts.FieldType):
-            dtype = dace_utils.as_dace_type(symbol_type.dtype)
+        elif isinstance(gt_type, ts.FieldType):
+            dtype = dace_utils.as_dace_type(gt_type.dtype)
             # use symbolic shape, which allows to invoke the program with fields of different size;
             # and symbolic strides, which enables decoupling the memory layout from generated code.
-            sym_shape, sym_strides = self._make_array_shape_and_strides(name, symbol_type.dims)
+            sym_shape, sym_strides = self._make_array_shape_and_strides(name, gt_type.dims)
             sdfg.add_array(name, sym_shape, dtype, strides=sym_strides, transient=transient)
 
-            return [(name, symbol_type)]
+            return [(name, gt_type)]
 
-        elif isinstance(symbol_type, ts.ScalarType):
-            dtype = dace_utils.as_dace_type(symbol_type)
-            # Scalar arguments passed to the program are represented as symbols in DaCe SDFG;
-            # the exception are members of tuple arguments, that are represented as scalar containers.
-            # The field size is sometimes passed as scalar argument to the program, so we have to
-            # check if the shape symbol was already allocated by `_make_array_shape_and_strides`.
-            # We assume that the scalar argument for field size always follows the field argument.
-            if is_tuple_member:
-                sdfg.add_scalar(name, dtype, transient=transient)
-            elif name in sdfg.symbols:
-                assert sdfg.symbols[name].dtype == dtype
-            else:
+        elif isinstance(gt_type, ts.ScalarType):
+            dtype = dace_utils.as_dace_type(gt_type)
+            if name in known_symbols:
                 sdfg.add_symbol(name, dtype)
+            elif dace_utils.is_field_symbol(name):
```

Comment:
What is the difference with the `if` branch just above.
From the descriptions (comment below and the definition of `known_symbols`) I would say that both branches address the same situation.

Reply:
They address the same situation but in a slightly different context.

If you take the gt4py unit tests, the range of cartesian domain is typically defined as [0, size].
Sometimes, when the field domain is implicitly derived from the field domain, the gt4py lowering adds the field size. Suppose a field `__sym`, then gt4py will add `__sym_size`. It never does this for the strides.

`program_sdfg(__sym1, __sym1_size_0, size, __sym1_stride_0)`

We are using `known_symbols` for the symbols used in domain expressions. The array symbols are derived from the field name (`__sym1` -> `__sym1_size_0, __sym1_stride_0`)

This if handles the specific case where gt4py adds the scalar argument for `__sym1_size_0`.

Reply:
Thanks for the explanation.
I know that behaviour but I did not associate it with it.
However, I think it does this only in very specific cases, Till once explained this to me.

Optionally: Ask Till what the conditions are that this happens and note them down here and also make a link to it.